src/Authentication/Authentication/custom/common/Permissions.ps1

```diff
@@ -1,36 +1,47 @@
 # ------------------------------------------------------------------------------
 #  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
 # ------------------------------------------------------------------------------
-
-
-$permissionsPulledFromMgGraphRequest = $null
+Update-FormatData -PrependPath $PSScriptRoot\..\..\Microsoft.Graph.Authentication.format.ps1xml
 
 function Permissions_GetPermissionsData {
+    param (
+        [bool] $online
+    )
 
     $permissions_MsGraphServicePrincipal = $null
+    $requestError = $null
+    $fromInvokeMgGraphRequest = $false
 
     # 2. Making a REST request to MS Graph
-    if ($null -eq $permissions_MsGraphServicePrincipal){
+
+    if (($null -eq $permissions_MsGraphServicePrincipal) -or ($null -ne $permissions_MsGraphServicePrincipal -and $fromInvokeMgGraphRequest -eq $false)) {
         $script:permissions_MsGraphServicePrincipal = try {
 
             # Write-Host "Getting data from web service"
             $result = Invoke-MgGraphRequest -method GET 'https://graph.microsoft.com/v1.0/servicePrincipals?filter=appId eq ''00000003-0000-0000-c000-000000000000''' 
             
             if ($null -ne $result) {
                 $result | select-object -expandproperty value 
+                $fromInvokeMgGraphRequest = $true
             }
 
         } catch [System.Management.Automation.ValidationMetadataException] {
 
-            # Write-Host "Getting data from local file"
+            $requestError = "Connection to MgGraph required. This can be done by using 'Connect-MgGraph'"
```

Comment:
Is this necessary? We could just record the error rather than using a custom string.. Also, it's not clear if we need to support localization. If we're going to add custom messages, we will need to plug in to localization framework. Might be something that can be done in a future PR.